pkg/webhook/bd_replacement.go

```diff
@@ -0,0 +1,402 @@
+	newSpec []apis.PoolSpec
+}
+
+// ValidateForBDReplacementCase validates the changes in CSPC for block device replacement in a raid group only if the
+// update/edit of CSPC can trigger a block device replacement.
+func ValidateForBDReplacementCase(commonPoolSpecs *poolspecs, bdr *BlockDeviceReplacement) (bool, string) {
+	for i, oldPoolSpec := range commonPoolSpecs.oldSpec {
+		oldPoolSpec := oldPoolSpec
+		if ok, msg := bdr.IsReplacementValid(&oldPoolSpec, &commonPoolSpecs.newSpec[i]); !ok {
+			return false, msg
+		}
+	}
+	return true, ""
+}
+
+// getCommonPoolSpecs get the same pool specs from old persisted CSPC and the new CSPC after modification
+// which is not persisted yet.
+func getCommonPoolSpecs(cspcNew, cspcOld *apis.CStorPoolCluster) (*poolspecs, error) {
+	commonPoolSpecs := &poolspecs{
+		oldSpec: []apis.PoolSpec{},
+		newSpec: []apis.PoolSpec{},
+	}
+	for _, oldPool := range cspcOld.Spec.Pools {
+		oldNodeName, err := nodeselect.GetNodeFromLabelSelector(oldPool.NodeSelector)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, newPool := range cspcNew.Spec.Pools {
+			newNodeName, err := nodeselect.GetNodeFromLabelSelector(oldPool.NodeSelector)
+			if err != nil {
+				return nil, err
+			}
+			if oldNodeName == newNodeName {
+				commonPoolSpecs.oldSpec = append(commonPoolSpecs.oldSpec, oldPool)
+				commonPoolSpecs.newSpec = append(commonPoolSpecs.newSpec, newPool)
+				break
+			}
+		}
+	}
+	return commonPoolSpecs, nil
+}
+
+// IsReplacementValid validates the pool specs on CSPC for block device replacement case.
+func (bdr *BlockDeviceReplacement) IsReplacementValid(oldPoolSpec, newPoolSpec *apis.PoolSpec) (bool, string) {
+	newToOldBd := make(map[string]string)
+	for _, oldRg := range oldPoolSpec.RaidGroups {
+		for _, newRg := range newPoolSpec.RaidGroups {
+			if IsRaidGroupCommon(oldRg, newRg) {
+				if IsBlockDeviceReplacementCase(oldRg, newRg) {
+					if ok, msg := bdr.IsBDReplacementValid(newRg, oldRg); !ok {
+						return false, msg
+					}
+					newBD := GetNewBDFromRaidGroups(newRg, oldRg)
+					for k, v := range newBD {
+						newToOldBd[k] = v
+					}
+					break
+				}
+			}
+		}
+	}
+
+	for newBD, oldBD := range newToOldBd {
+		err := createBDC(newBD, oldBD, bdr.OldCSPC)
+		if err != nil {
+			return false, err.Error()
+		}
+	}
+	return true, ""
+}
+
+// IsRaidGroupCommon returns true if the provided raid groups are the same raid groups.
+func IsRaidGroupCommon(rgOld, rgNew apis.RaidGroup) bool {
+	oldBdMap := make(map[string]bool)
+	for _, bd := range rgOld.BlockDevices {
+		oldBdMap[bd.BlockDeviceName] = true
+	}
+
+	for _, bd := range rgNew.BlockDevices {
+		if oldBdMap[bd.BlockDeviceName] {
+			return true
+		}
+	}
+	return false
+}
+
+// IsBlockDeviceReplacementCase returns true if the edit/update of CSPC can trigger a blockdevice
+// replacement.
+func IsBlockDeviceReplacementCase(newRaidGroup, oldRaidGroup apis.RaidGroup) bool {
+	if oldRaidGroup.Type == string(apis.PoolStriped) {
```

Comment:
some times the raid group can be empty I think at this time we have to pick from pool config.